aws/resource_aws_opsworks_chef.go

```diff
@@ -0,0 +1,335 @@
+	"github.com/aws/aws-sdk-go/service/opsworkscm"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+const EngineModel = "Single"
+const EngineVersion = "12"
+
+func resourceAwsOpsworksChef() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsOpsworksChefCreate,
+		Read:   resourceAwsOpsworksChefRead,
+		Update: resourceAwsOpsworksChefUpdate,
+		Delete: resourceAwsOpsworksChefDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			// TODO: do we want this? I think this is another "magic" thing that happens
+			// I think we can still associate a public IP later if we want? :shrug:
+			"associate_public_ip_address": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+
+			"backup_retention_count": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default:  1,
+			},
+
+			"disable_automated_backup": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+
+			// TODO:
+			// I kinda want to make this required because how else are we going to get at that info?
+			// Also, since these are write-only and not returned by the API, we'll probably need a custom diff function that ignores them?
+			"chef_pivotal_key": {
+				Type:      schema.TypeString,
+				Optional:  true,
+				Default:   nil,
+				Sensitive: true,
+			},
+
+			"chef_delivery_admin_password": {
+				Type:      schema.TypeString,
+				Optional:  true,
+				Default:   nil,
+				Sensitive: true,
+			},
+
+			// TODO:
+			// default and only valid. I'll leave it configurable for future proofing but
+			// validation failure, maybe?
+			"engine_model": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  EngineModel,
+			},
+
+			// TODO:
+			// default and only valid. I'll leave it configurable for future proofing but
+			// validation failure, maybe?
+			"engine_version": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  EngineVersion,
+			},
+
+			// TODO: document the instance role required
+			// https://s3.amazonaws.com/opsworks-cm-us-east-1-prod-default-assets/misc/opsworks-cm-roles.yaml
+			"instance_profile_arn": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"instance_type": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"ssh_key_pair": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"preferred_backup_window": {
+				Type:     schema.TypeString,
+				Required: true, // TODO: ?
+			},
+
+			"preferred_maintenance_window": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			// TODO:
+			// if you don't specify this, it'll create one. I'm of the opinion that users should
+			// create it themselves. Or maybe some sort of wrapper module? I dunno.
+			// if it's optional, will they be computed? if so, is that something we can express here?
```

Comment:
Here's another "magic happens" thing. If you don't specify security groups, it'll make one for you. Is this something we want to allow or should we require folks to specify the security groups explicitly? I'm leaning on the side of explicitly.